uspace/lib/cpp/include/__bits/thread/async.hpp

```diff
@@ -0,0 +1,146 @@
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef LIBCPP_BITS_THREAD_ASYNC
+#define LIBCPP_BITS_THREAD_ASYNC
+
+#include <__bits/thread/future.hpp>
+#include <__bits/thread/future_common.hpp>
+#include <__bits/thread/shared_state.hpp>
+#include <__bits/type_traits/result_of.hpp>
+#include <__bits/utility/forward_move.hpp>
+
+namespace std
+{
+
+    enum class launch
+    {
+        async = 1,
+        deferred
+    };
+
+    namespace aux
+    {
+        /**
+         * Note: The reason we keep the actual function
+         *       within the aux namespace is that were the non-policy
+         *       version of the function call the other one in the std
+         *       namespace, we'd get resolution conflicts. This way
+         *       aux::async is properly called even if std::async is
+         *       called either with or without a launch policy.
+         */
+        template<class F, class... Args>
+        future<result_of_t<decay_t<F>(decay_t<Args>...)>>
+        async(launch policy, F&& f, Args&&... args)
+        {
+            using result_t = result_of_t<decay_t<F>(decay_t<Args>...)>;
+
+            bool async = (static_cast<int>(policy) &
+                          static_cast<int>(launch::async)) != 0;
+            bool deferred = (static_cast<int>(policy) &
+                             static_cast<int>(launch::deferred)) != 0;
+
+            /**
+             * Note: The case when async | deferred is set in policy
+             *       is implementation defined, feel free to change.
+             */
+            if (async && deferred)
+            {
+                return future<result_t>{
+                    new aux::deferred_shared_state<
+                        result_t, F, Args...
+                    >{forward<F>(f), forward<Args>(args)...}
+                };
+            }
+            else if (async)
+            {
+                return future<result_t>{
+                    new aux::async_shared_state<
+                        result_t, F, Args...
+                    >{forward<F>(f), forward<Args>(args)...}
+                };
+            }
+            else if (deferred)
+            {
+               /**
+                * Duplicated on purpose because of the default.
+                * Do not remove!
+                */
+                return future<result_t>{
+                    new aux::deferred_shared_state<
+                        result_t, F, Args...
+                    >{forward<F>(f), forward<Args>(args)...}
+                };
+            }
+
+            /**
+             * This is undefined behaviour, let's be nice though ;)
```

Comment:
If the spec says it's undefined behavior, and we detect it, it should abort the program. Allowing undefined behavior to go undetected "to be nice" is a terrible idea. In the case of C++, throwing `std::invalid_argument` is probably the way to go?

Reply:
> Allowing undefined behavior to go undetected "to be nice" is a terrible idea.

Is it? I am not aware of other implementations failing/crashing/reporting when they encounter undefined behaviour. The libstdc++ implementation on my tumbleweed does not crash, it asynchronously executes the given functor.

Since what is visible from the outside should be and is the same for all policies that I'm aware of, I think that falling back to a default here is a pretty reasonable thing to do (the reason for the deferred policy being the default can be read in the other commend - basically no risk of surprises).

Reply:
Is it really undefined behavior? In C and C++, "undefined behavior" has very specific meaning. It means that any program that touches it loses all semantic meaning. It can start WW III. It can make demons fly out your nose. It can execute the functor as usual. It can do one at those at random in every run. The point is, undefined behavior is the ultimate bad and if you already detect it, there is no reason ever not to make it crash loudly.

Reply:
If you think that's overreacting, just remember that GCC will be happy to just arbitrary silently remove your code without telling you, whenever it detects that it would invoke undefined behavior. -_-

Reply:
Actually, the definition of undefined behaviour is (1.3.24, defns.undefined):

> behaviour for which this International Standard imposes no requirements

The standard explicitly allows (among others):

* ignoring the situation completely
* behaving during translation or execution in a documented manner (with or without a message)
* terminating translation or execution (with a message)

Any of these are usable, but since - as I previously noted - the outside effect of the policy should not be visible to the user and libstdc++ itself does not fail in this case, I think that we should keep the deferred default here (plus it helps compatibility with other implementations).

Reply:
"imposes no requirements" means what I said. It can have arbitrary nonlocal effects. Compiler optimizers heavily utilize any such lack of requirements by assuming "undefined behavior can't happen". Any program that involves undefined behavior in any way is implicitly buggy in unforeseeable ways and should not be accepted. Compatibility doesn't play a role because where undefined behavior is concerned, the program doesn't have any valid meaning.

Reply:
Well, consider the following scenario:

Implementation X adds std::launch::thread_pool, which when used with std::async causes std::async to spawn the asynchronous task in a predefined thread pool as an optimization. The behaviour will be the same (sans the optimization) for us given the semantics of std::async and it is actually _defined behaviour_ for that implementation - this is because the standard actually says that undefined behaviour here is triggered if the launch policy is not defined by the standard _and_ not defined by the implementation (so it can be undefined behaviour on some implementation while defined behaviour on others).

If we leave it as it is, program using std::async::thread_pool written for implementation X and ported over will still work as it did before (sans the optimization of course).